src/app-layer-ssl.c

```diff
@@ -255,6 +259,249 @@ static void SSLSetTxDetectFlags(void *vtx, uint8_t dir, uint64_t flags)
+        return 0;
+    }
+
+    ssl_state->server_connp.cert0_issuerdn = SCStrdup(buffer);
+    if (ssl_state->server_connp.cert0_issuerdn == NULL)
+        return -1;
+
+    return 0;
+}
+
+static inline int TlsDecodeHSCertificateSerial(SSLState *ssl_state,
+                                               Asn1Generic *cert)
+{
+    if (unlikely(ssl_state->server_connp.cert0_serial != NULL))
+        return 0;
+
+    uint32_t err = 0;
+    char buffer[512];
+
+    int rc = Asn1DerGetSerial(cert, buffer, sizeof(buffer), &err);
+    if (rc != 0) {
+        TlsDecodeHSCertificateErrSetEvent(ssl_state, err);
+        return 0;
+    }
+
+    ssl_state->server_connp.cert0_serial = SCStrdup(buffer);
+    if (ssl_state->server_connp.cert0_serial == NULL)
+        return -1;
+
+    return 0;
+}
+
+static inline int TlsDecodeHSCertificateValidity(SSLState *ssl_state,
+                                                 Asn1Generic *cert)
+{
+    uint32_t err = 0;
+    time_t not_before;
+    time_t not_after;
+
+    int rc = Asn1DerGetValidity(cert, &not_before, &not_after, &err);
+    if (rc != 0) {
+        TlsDecodeHSCertificateErrSetEvent(ssl_state, err);
+        return 0;
+    }
+
+    ssl_state->server_connp.cert0_not_before = not_before;
+    ssl_state->server_connp.cert0_not_after = not_after;
+
+    return 0;
+}
+
+static inline int TlsDecodeHSCertificateFingerprint(SSLState *ssl_state,
+                                                    uint8_t *input,
+                                                    uint32_t cert_len)
+{
+    if (unlikely(ssl_state->server_connp.cert0_fingerprint != NULL))
+        return 0;
+
+    ssl_state->server_connp.cert0_fingerprint = SCMalloc(SHA1_STRING_LENGTH *
+                                                         sizeof(char));
+    if (ssl_state->server_connp.cert0_fingerprint == NULL)
+        return -1;
+
+    uint8_t *hash = ComputeSHA1(input, cert_len);
+    if (hash == NULL)
+        return 0;
+
+    int i, x;
+    for (i = 0, x = 0; x < SHA1_LENGTH; x++)
+    {
+        i += snprintf(ssl_state->server_connp.cert0_fingerprint + i,
+                      SHA1_STRING_LENGTH - i, i == 0 ? "%02x" : ":%02x",
+                      *(hash + x));
+    }
+
+    SCFree(hash);
+
+    return 0;
+}
+
+static inline int TlsDecodeHSCertificateAddCertToChain(SSLState *ssl_state,
+                                                       uint8_t *input,
+                                                       uint32_t cert_len)
+{
+    SSLCertsChain *cert = SCCalloc(1, sizeof(SSLCertsChain));
+    if (cert == NULL)
+        return -1;
+
+    cert->cert_data = input;
+    cert->cert_len = cert_len;
+    TAILQ_INSERT_TAIL(&ssl_state->server_connp.certs, cert, next);
+
+    return 0;
+}
+
+static int TlsDecodeHSCertificate(SSLState *ssl_state,
+                                  const uint8_t * const initial_input,
+                                  const uint32_t input_len)
+{
+    uint8_t *input = (uint8_t *)initial_input;
```

Comment:
this can be const too right? The ptr itself can change, but the content can't?

Reply:
Sure, just have to typecast a couple of places, but yeah, the content can't be changed.